Replace add_spawn() calls in mapgen with place_spawn()

There are several building layouts in mapgen.cpp and building_generation.cpp that have add_spawn() calls placing specific monsters on the map instead of place_spawn, which is more generic and selects monsters from a group.  This causes issues with the options that modify spawn behavior, because add_spawn() doesn't support those options.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/1366883-replace-add_spawn-calls-in-mapgen-with-place_spawn?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

Making it private:
```diff
diff --git a/src/map.h b/src/map.h
index 17dbf6727d..01f8086223 100644
--- a/src/map.h
+++ b/src/map.h
@@ -1719,13 +1719,7 @@ class map
         // places an NPC, if static NPCs are enabled or if force is true
         character_id place_npc( const point &p, const string_id<npc_template> &type );
         void apply_faction_ownership( const point &p1, const point &p2, const faction_id &id );
-        void add_spawn( const mtype_id &type, int count, const tripoint &p,
-                        bool friendly = false, int faction_id = -1, int mission_id = -1,
-                        const std::string &name = "NONE" );
-        void add_spawn( const mtype_id &type, int count, const tripoint &p, bool friendly,
-                        int faction_id, int mission_id, const std::string &name,
-                        const spawn_data &data );
-        void add_spawn( const MonsterGroupResult &spawn_details, const tripoint &p );
+
         void do_vehicle_caching( int z );
         // Note: in 3D mode, will actually build caches on ALL z-levels
         void build_map_cache( int zlev, bool skip_lightmap = false );
@@ -1873,6 +1867,15 @@ class map
         */
         void rotten_item_spawn( const item &item, const tripoint &p );
     private:
+       // Use place_spawns or JSON mapgen instead
+        void add_spawn( const mtype_id &type, int count, const tripoint &p,
+                        bool friendly = false, int faction_id = -1, int mission_id = -1,
+                        const std::string &name = "NONE" );
+        void add_spawn( const mtype_id &type, int count, const tripoint &p, bool friendly,
+                        int faction_id, int mission_id, const std::string &name,
+                        const spawn_data &data );
+        void add_spawn( const MonsterGroupResult &spawn_details, const tripoint &p );
+
         // Helper #1 - spawns monsters on one submap
         void spawn_monsters_submap( const tripoint &gp, bool ignore_sight, bool spawn_nonlocal = false );
         // Helper #2 - spawns monsters on one submap and from one group on this submap
```

There are three calls outside of `map`:
Fields spawning monsters?
```sh
src/map_field.cpp: In function ‘void field_processor_monster_spawn(const tripoint&, field_entry&, field_proc_data&)’:
src/map_field.cpp:768:38: error: ‘void map::add_spawn(const MonsterGroupResult&, const tripoint&)’ is private within this context
  768 |                     pd.here.add_spawn( mgr, *spawn_point );
      |                     ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
```
That map extra
```sh
src/map_extras.cpp: In function ‘bool MapExtras::mx_fungal_zone(map&, const tripoint&)’:
src/map_extras.cpp:2168:25: error: ‘void map::add_spawn(const mtype_id&, int, const tripoint&, bool, int, int, const std::string&)’ is private within this context
 2168 |     fungal_map.add_spawn( mon_fungaloid_queen, 1, suitable_location );
      |     ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Monster reproduction
```sh
src/monster.cpp: In member function ‘void monster::try_reproduce()’:
src/monster.cpp:607:31: error: ‘void map::add_spawn(const mtype_id&, int, const tripoint&, bool, int, int, const std::string&)’ is private within this context
  607 |                 here.add_spawn( type->baby_monster, spawn_cnt, pos() );
      |                 ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
JSON Mapgen:
```sh
src/mapgen.cpp: In member function ‘virtual void jmapgen_monster::apply(const mapgendata&, const jmapgen_int&, const jmapgen_int&, const std::string&) const’:
src/mapgen.cpp:2458:36: error: ‘void map::add_spawn(const mtype_id&, int, const tripoint&, bool, int, int, const std::string&, const spawn_data&)’ is private within this context
 2458 |                     dat.m.add_spawn( mgr.name, spawn_count * pack_size.get(),
      |                     ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 2459 |                     { x.get(), y.get(), dat.m.get_abs_sub().z() },
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 2460 |                     friendly, -1, mission_id, chosen_name, data );
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
The map extra is trivial, the rest probably aren't.

--- a/data/json/monstergroups/fungi.json
+++ b/data/json/monstergroups/fungi.json
@@ -72,5 +72,10 @@
       { "monster": "mon_zombie_fungus_acidic", "weight": 50, "cost_multiplier": 0 },
       { "monster": "mon_zombie_dog_fungus", "weight": 50, "cost_multiplier": 0 }
     ]
+  },
+  {
+    "type": "monstergroup",
+    "id": "GROUP_FUNGI_QUEEN",
+    "monsters": [ { "monster": "mon_fungaloid_queen", "weight": 1000, "cost_multiplier": 0 } ]
   }
 ]

--- a/src/map.h
+++ b/src/map.h
@@ -386,6 +386,13 @@ class map
                 field_proc_data & );
         friend void field_processor_fd_incendiary( const tripoint_bub_ms &, field_entry &,
                 field_proc_data & );
+        friend void field_processor_monster_spawn( const tripoint_bub_ms &, field_entry &,
+                field_proc_data & );
+
+        // add_spawn is private; these classes need it for data-driven spawning
+        // (JSON mapgen and monster reproduction). Mapgen code should use place_spawns() instead.
+        friend class jmapgen_monster;
+        friend class monster;
 
         // for testing
         friend class map_meddler;
@@ -1734,13 +1741,6 @@ class map
         character_id place_npc( const point_bub_ms &p, const string_id<npc_template> &type );
         void apply_faction_ownership( const point_bub_ms &p1, const point_bub_ms &p2,
                                       const faction_id &id );
-        void add_spawn( const mtype_id &type, int count, const tripoint_bub_ms &p,
-                        bool friendly = false, int faction_id = -1, int mission_id = -1,
-                        const std::optional<std::string> &name = std::nullopt );
-        void add_spawn( const mtype_id &type, int count, const tripoint_bub_ms &p, bool friendly,
-                        int faction_id, int mission_id, const std::optional<std::string> &name,
-                        const spawn_data &data );
-        void add_spawn( const MonsterGroupResult &spawn_details, const tripoint_bub_ms &p );
         void do_vehicle_caching( int z );
         // Note: in 3D mode, will actually build caches on ALL z-levels
         void build_map_cache( int zlev, bool skip_lightmap = false );
@@ -1876,6 +1876,14 @@ class map
         */
         void rotten_item_spawn( const item &item, const tripoint_bub_ms &p );
     private:
+        void add_spawn( const mtype_id &type, int count, const tripoint_bub_ms &p,
+                        bool friendly = false, int faction_id = -1, int mission_id = -1,
+                        const std::optional<std::string> &name = std::nullopt );
+        void add_spawn( const mtype_id &type, int count, const tripoint_bub_ms &p, bool friendly,
+                        int faction_id, int mission_id, const std::optional<std::string> &name,
+                        const spawn_data &data );
+        void add_spawn( const MonsterGroupResult &spawn_details, const tripoint_bub_ms &p );
+
         // Helper #1 - spawns monsters on one submap
         void spawn_monsters_submap( const tripoint_rel_sm &gp, bool ignore_sight,
                                     bool spawn_nonlocal = false );
@@ -2379,12 +2387,6 @@ class tinymap : private map
                                friendly,
                                name, mission_id );
         }
-        void add_spawn( const mtype_id &type, int count, const tripoint_omt_ms &p,
-                        bool friendly = false, int faction_id = -1, int mission_id = -1,
-                        const std::optional<std::string> &name = std::nullopt ) {
-            map::add_spawn( type, count, rebase_bub( p ), friendly, faction_id, mission_id, name );
-        }
-
         using map::translate;
         ter_id ter( const tripoint_omt_ms &p ) const {
             return map::ter( rebase_bub( p ) );

--- a/src/map_extras.cpp
+++ b/src/map_extras.cpp
@@ -93,13 +93,13 @@ static const map_extra_id map_extra_mx_shrubbery( "mx_shrubbery" );
 
 static const mongroup_id GROUP_FISH( "GROUP_FISH" );
 static const mongroup_id GROUP_FUNGI_FUNGALOID( "GROUP_FUNGI_FUNGALOID" );
+static const mongroup_id GROUP_FUNGI_QUEEN( "GROUP_FUNGI_QUEEN" );
 static const mongroup_id GROUP_MIL_PASSENGER( "GROUP_MIL_PASSENGER" );
 static const mongroup_id GROUP_MIL_PILOT( "GROUP_MIL_PILOT" );
 static const mongroup_id GROUP_MIL_WEAK( "GROUP_MIL_WEAK" );
 static const mongroup_id GROUP_NETHER_PORTAL( "GROUP_NETHER_PORTAL" );
 static const mongroup_id GROUP_STRAY_DOGS( "GROUP_STRAY_DOGS" );
 
-static const mtype_id mon_fungaloid_queen( "mon_fungaloid_queen" );
 
 static const oter_type_str_id oter_type_road( "road" );
 
@@ -1049,7 +1049,8 @@ static bool mx_fungal_zone( map &m, const tripoint_abs_sm &abs_sub )
     }
 
     const tripoint_bub_ms suitable_location = random_entry( suitable_locations, omt_center );
-    m.add_spawn( mon_fungaloid_queen, 1, suitable_location );
+    m.place_spawns( GROUP_FUNGI_QUEEN, 1, suitable_location.xy(), suitable_location.xy(),
+                    suitable_location.z(), 1, true );
     m.place_spawns( GROUP_FUNGI_FUNGALOID, 1,
                     suitable_location.xy() + point::north_west,
                     suitable_location.xy() + point::south_east,

--- a/src/map_field.cpp
+++ b/src/map_field.cpp
@@ -760,8 +760,8 @@ static void field_processor_fd_electricity( const tripoint_bub_ms &p, field_entr
     }
 }
 
-static void field_processor_monster_spawn( const tripoint_bub_ms &p, field_entry &cur,
-        field_proc_data &pd )
+void field_processor_monster_spawn( const tripoint_bub_ms &p, field_entry &cur,
+                                    field_proc_data &pd )
 {
     const field_intensity_level &int_level = cur.get_intensity_level();
     int monster_spawn_chance = int_level.monster_spawn_chance;

